### PR TITLE
cmd: add genesis-gen tool

### PR DESCRIPTION
**Pull request by 0xcb9ff9**
### Description

related issue #2  

Add `genesis-gen` tool, helper create genesis.json file

### Example

```sh
genesis-gen --chain-id 2333 \
    --epoch-size 300 \
    --validator 0xC035596a28ed3C5E357baf2860baCB6d665E81d5 \
    --premine 0x23d7B7be9B63e63F2e76b39bEA23BcD517731a0b:0x3635C9ADC5DEA0000000 \
    --validatorset-owner 0x20E088469849516610a759C7Bd60434298a91589
```

### Changes

Notable changes: 
* use `go:generate` pre-convert the contract hex string to a binary file
* contract byte code use `go:embed` embed binary file, avoid runtime conversion
* add `genesis-gen` tool

### CODE REVIEW

File: cmd/genesis-gen/main.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+		log.Warn("no bridge signers specified")
+	}
+
+	chainConfig := &params.ChainConfig{
+		ChainID:             new(big.Int).SetUint64(c.GlobalUint64(chainIDFlag.Name)),
+		HomesteadBlock:      big.NewInt(0),
+		EIP150Block:         big.NewInt(0),
+		EIP155Block:         big.NewInt(0),
+		EIP158Block:         big.NewInt(0),
+		IBFTBlock:           nil, // this is dogechain v1 network block, new network should not use this
```
Comment (DarianShawn) [inline cmd/genesis-gen/main.go]: `IBFTBlock` should be enable on genesis block.

File: cmd/genesis-gen/main.go [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+		ByzantiumBlock:      big.NewInt(0),
+		ConstantinopleBlock: big.NewInt(0),
+		PetersburgBlock:     big.NewInt(0),
+		IstanbulBlock:       big.NewInt(0),
+		PreportlandBlock:    big.NewInt(0),
```
Comment (DarianShawn) [inline cmd/genesis-gen/main.go]: `preportland, portland, detroit`, these hard forks need to be upgraded, so we should not provide it from beginning.
Or, we should set those blocks with different block numbers.

File: cmd/genesis-gen/main.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+		IBFT: &params.IBFTConfig{
+			EpochSize: c.GlobalUint64(epochSizeFlag.Name),
+			Type:      params.IBFTPoS,
+		},
+	}
+
+	genesis := &core.Genesis{
+		Config:     chainConfig,
+		Nonce:      0,
+		Timestamp:  uint64(time.Now().Unix()),
```
Comment (DarianShawn) [inline cmd/genesis-gen/main.go]: The timestamp should be configurable? Or, `0` if you want it Mainnet or Devnet compatible.

File: core/dccontracts/upgrader.go [hunk 4]
```diff
@@ -70,7 +69,7 @@ func init() {
...
...0052601160045260246000fd5b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601260045260246000fd5b7f4e487b7100000000000000000000000000000000000000000000000000000000600052602260045260246000fd5b7f4e487b7100000000000000000000000000000000000000000000000000000000600052603160045260246000fd5b7f4e487b7100000000000000000000000000000000000000000000000000000000600052603260045260246000fd5b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b600080fd5b600080fd5b600080fd5b600080fd5b6000601f19601f8301169050919050565b60008160601b9050919050565b7f4f776e61626c653a206e6577206f776e657220697320746865207a65726f206160008201527f6464726573730000000000000000000000000000000000000000000000000000602082015250565b7f4f6e6c79206f776e65722063616e2063616c6c2066756e6374696f6e00000000600082015250565b7f466f726269640000000000000000000000000000000000000000000000000000600082015250565b7f4f6e6c79207369676e65722063616e2063616c6c2066756e6374696f6e000000600082015250565b61230681611f83565b811461231157600080fd5b50565b61231d81611fc1565b811461232857600080fd5b5056fea26469706673582212208879968f26c5ad21d7e3b58081ca73de324deb7256e1760798788f290bffec8364736f6c63430008060033",
```
Comment (DarianShawn) [inline core/dccontracts/upgrader.go:73]: Why do you delete our upgraded codes?
The byte codes are totally different after some hard forks.
I think you miss out the most important part.

**Review by DarianShawn — COMMENTED** [review]
Great to sea this tool.
But I'm afraid that some thing bad happen here...

File: cmd/genesis-gen/main.go (hunk 1 above)
Comment (0xcb9ff9) [reply cmd/genesis-gen/main.go]: ok, done [94cf2a7](https://github.com/dogechain-lab/dbsc/pull/17/commits/94cf2a7ecaedb45a2b0df31664a99e8964528c64)

File: cmd/genesis-gen/main.go (hunk 2 above)
Comment (0xcb9ff9) [reply cmd/genesis-gen/main.go]: default configure `preportland, portland, detroit` upgrade in block 1,2,3 ? [94cf2a7](https://github.com/dogechain-lab/dbsc/pull/17/commits/94cf2a7ecaedb45a2b0df31664a99e8964528c64)

File: core/dccontracts/upgrader.go (hunk 4 above)
Comment (0xcb9ff9) [reply core/dccontracts/upgrader.go:73]: My mistake, I misunderstood the process of upgrading the contract and initialization, now use `go:generate` convert the contract hex string to a binary file, and generate `go:embed` code to embed binary file in the execution file

File: cmd/genesis-gen/main.go (hunk 3 above)
Comment (0xcb9ff9) [reply cmd/genesis-gen/main.go]: [39cfd61](https://github.com/dogechain-lab/dbsc/pull/17/commits/39cfd616ccee8b5e388dfd2f3678569f9e48415c) fix

**Review by DarianShawn — APPROVED** [review]
LGTM
I've checked with the byte codes and shasum and find all match.

**Pull request merged**